src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,482 @@
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+void
+PerfLogImp::jobQueue(std::underlying_type_t<JobType> const type)
+{
+    auto counter = counters_.jq_.find(type);
+    if (counter != counters_.jq_.end())
+    {
+        std::lock_guard<std::mutex> lock(counter->second.mut);
+        ++counter->second.queued;
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+void
+PerfLogImp::jobStart(std::underlying_type_t<JobType> const type,
+    std::uint64_t dur,
+    std::chrono::time_point<std::chrono::steady_clock> startTime,
+    int instance)
+{
+    auto counter = counters_.jq_.find(type);
+    if (counter != counters_.jq_.end())
+    {
+        {
+            std::lock_guard<std::mutex> lock(counter->second.mut);
+            ++counter->second.started;
+            counter->second.queuedDuration += dur;
+        }
+        std::lock_guard<std::mutex> lock(counters_.jobsMutex_);
+        if (instance < counters_.jobs_.size())
+            counters_.jobs_[instance] = std::make_pair(type, startTime);
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+void
+PerfLogImp::jobFinish(std::underlying_type_t<JobType> const type,
+    std::uint64_t dur, int instance)
+{
+    auto counter = counters_.jq_.find(type);
+    if (counter != counters_.jq_.end())
+    {
+        {
+            std::lock_guard<std::mutex> lock(counter->second.mut);
+            ++counter->second.finished;
+            counter->second.runningDuration += dur;
+        }
+        std::lock_guard<std::mutex> lock(counters_.jobsMutex_);
+        if (instance < counters_.jobs_.size())
+            counters_.jobs_[instance] = std::make_pair(jtINVALID,
+                std::chrono::time_point<std::chrono::steady_clock>());
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+void
+PerfLogImp::rpcError(std::string const& method, std::uint64_t requestId)
+{
+    auto counter = counters_.rpc_.find(method);
+    if (counter != counters_.rpc_.end())
+    {
+        std::chrono::time_point<std::chrono::steady_clock> startTime;
+        {
+            std::lock_guard<std::mutex> lock(counters_.methodsMutex_);
+            auto const e = counters_.methods_.find(requestId);
+            if (e != counters_.methods_.end())
+            {
+                startTime = e->second.second;
+                counters_.methods_.erase(e);
+            }
+            else
+            {
+                assert(false);
+            }
+        }
+        std::lock_guard<std::mutex> lock(counter->second.mut);
+        ++counter->second.errored;
+        counter->second.duration +=
+            std::chrono::duration_cast<std::chrono::microseconds>(
+                startTime - std::chrono::steady_clock::now()).count();
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+void
+PerfLogImp::resizeJobs(int const resize)
+{
+    std::lock_guard<std::mutex> lock(counters_.jobsMutex_);
+    counters_.workers_ = resize;
+    if (resize > counters_.jobs_.size())
```

Comment:
what if size goes down - would that be considered an error, or expected? If it shrinks but you don't shrink this collection, won't we report some invalid entries moving forward ?

Reply:
The actual number of running threads exit asynchronously. If the vector shrinks then the running jobs won't be reported when they exit. There should be no harm (other than a slight cost in walking the vector) if the vector is too big because if no jobs are running, they aren't printed (jtINVALID).